From 421d44884505e2aac5cfcb9edd1a8dae5fed687d Mon Sep 17 00:00:00 2001
From: Roel van Meer <roel@1afa.com>
Date: Wed, 7 Sep 2022 22:01:33 +0200
Subject: [PATCH 2/2] Copy timeperiod names into contact struct members

This fixes a problem where we'd copy the address instead of the content
of the timeperiod name to the host_notification_period and
service_notification_period members of contacts, causing a crash later
when they were free'd.

t-tap/test_commands now passes again.

Found while looking for the cause of #285, but doesn't seem to be
related.

Signed-off-by: Roel van Meer <roel@1afa.com>
---
 src/naemon/objects_contact.c | 4 ++--
 src/naemon/xrddefault.c      | 4 ++--
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/naemon/objects_contact.c b/src/naemon/objects_contact.c
index 072d6d322..b33b8d6a2 100644
--- a/src/naemon/objects_contact.c
+++ b/src/naemon/objects_contact.c
@@ -71,8 +71,8 @@ int setup_contact_variables(contact *new_contact, const char *alias, const char
 		return -1;
 	}
 
-	new_contact->host_notification_period = htp ? htp->name : NULL;
-	new_contact->service_notification_period = stp ? stp->name : NULL;
+	new_contact->host_notification_period = htp ? nm_strdup(htp->name) : NULL;
+	new_contact->service_notification_period = stp ? nm_strdup(stp->name) : NULL;
 	new_contact->host_notification_period_ptr = htp;
 	new_contact->service_notification_period_ptr = stp;
 	if (alias)
diff --git a/src/naemon/xrddefault.c b/src/naemon/xrddefault.c
index c24e74392..1dc88859d 100644
--- a/src/naemon/xrddefault.c
+++ b/src/naemon/xrddefault.c
@@ -1504,7 +1504,7 @@ int xrddefault_read_state_information(void)
 								/* make sure the timeperiod still exists... */
 								temp_timeperiod = find_timeperiod(val);
 								if (temp_timeperiod) {
-									temp_contact->host_notification_period = temp_timeperiod->name;
+									temp_contact->host_notification_period = nm_strdup(temp_timeperiod->name);
 									temp_contact->host_notification_period_ptr = temp_timeperiod;
 								} else {
 									temp_contact->modified_host_attributes &= ~MODATTR_NOTIFICATION_TIMEPERIOD;
@@ -1516,7 +1516,7 @@ int xrddefault_read_state_information(void)
 								/* make sure the timeperiod still exists... */
 								temp_timeperiod = find_timeperiod(val);
 								if (temp_timeperiod) {
-									temp_contact->service_notification_period = temp_timeperiod->name;
+									temp_contact->service_notification_period = nm_strdup(temp_timeperiod->name);
 									temp_contact->service_notification_period_ptr = temp_timeperiod;
 								} else {
 									temp_contact->modified_service_attributes &= ~MODATTR_NOTIFICATION_TIMEPERIOD;
